File: lexer.py
# coding=utf-8

import sys
import string
import keywords

def pretreatment(file_name):
    '''
    预处理函数，将文件中的空格，换行等无关字符处理掉
    :param file_name: 处理文件的路径
    :return:
    '''
    try:
        fp_read = open(file_name, 'r')
        fp_write = open('file.tmp', 'w')
        sign = 0  # 是否遇到字符（非空格）
        # 不断读取，进行预处理
        while True:
            read = fp_read.readline()
            if not read:  # 读取结束
                break
            length = len(read)
            i = -1  # 索引
            while i < length - 1:
                i += 1
                if sign == 0:  # 还没有遇到字符
                    if read[i] == ' ':
                        continue
                if read[i] == '#':  # 后面均为注释，直接结束此行分析
                    break
                elif read[i] == ' ':  #
                    if sign == 1:
                        continue
                    else:
                        sign = 1
                        fp_write.write(' ')
                elif read[i] == '\t':  # table处理和空格相同
                    if sign == 1:
                        continue
                    else:
                        sign = 1
                        fp_write.write(' ')
                elif read[i] == '\n':  # 换行
                    if sign == 1:
                        continue
                    else:
                        fp_write.write(' ')  #
                        sign = 1
                elif read[i] == '"':
                    fp_write.write(read[i])
                    i += 1
                    while i < length and read[i] != '"':
                        fp_write.write(read[i])
                        i += 1
                    if i >= length:
                        break
                    fp_write.write(read[i])
                elif read[i] == "'":
                    fp_write.write(read[i])
                    i += 1
                    while i < length and read[i] != "'":
                        fp_write.write(read[i])
                        i += 1
                    if i >= length:
                        break
                    fp_write.write(read[i])
                else:
                    sign = 3
                    fp_write.write(read[i])
    except Exception:
        print(file_name, ': This FileName Not Found!')


def save(string):
    '''

    :param string:
    :return:
    '''
    if string in keywords.keywords.keys():
        if string not in keywords.signlist.keys():
            keywords.signlist[string] = keywords.keywords[string]
    else:
        try:
            float(string)
            save_const(string)
        except ValueError:
            save_var(string)


def save_var(string):
    '''

    :param string:
    :return:
    '''
    if string not in keywords.signlist.keys():
        if len(string.strip()) < 1:
            pass
        else:
            if is_signal(string) == 1:
                keywords.signlist[string] = 301
            else:
                keywords.signlist[string] = 501


def save_const(string):
    '''

    :param string:
    :return:
    '''
    if string not in keywords.signlist.keys():
        keywords.signlist[string] = 401


def save_error(string):
    '''

    :param string:
    :return:
    '''
    if string not in keywords.signlist.keys():
        keywords.signlist[string] = 501


def is_signal(s):
    '''

    :param s:
    :return:
    '''
    if s[0] == '_' or s[0] in string.ascii_letters:
        for i in s:
            if i in string.ascii_letters or i == '_' or i in string.digits:
                pass
            else:
                return 0
        return 1
    else:
        return 0


def recognition(filename):
    """

    :param filename:
    :return:
    """
    try:
        fp_read = open(filename, 'r')
        string = ""
        sign = 0
        while True:
            read = fp_read.read(1)
            if not read:
                break

            if read == ' ':
                if len(string.strip()) < 1:
                    sign = 0
                    pass
                else:
                    if sign == 1 or sign == 2:
                        string += read
                    else:
                        save(string)
                        string = ""
                        sign = 0
            elif read == '(':
                if sign == 1 or sign == 2:
                    string += read
                else:
                    save(string)
                    string = ""
                    save('(')
            elif read == ')':
                if sign == 1 or sign == 2:
                    string += read
                else:
                    save(string)
                    string = ""
                    save(')')
            elif read == '[':
                if sign == 1 or sign == 2:
                    string += read
                else:
                    save(string)
                    string = ""
                    save('[')
            elif read == ']':
                if sign == 1 or sign == 2:
                    string += read
                else:
                    save(string)
                    string = ""
                    save(']')
            elif read == '{':
                if sign == 1 or sign == 2:
                    string += read
                else:
                    save(string)
                    string = ""
                    save('{')
            elif read == '}':
                if sign == 1 or sign == 2:
                    string += read
                else:
                    save(string)
                    string = ""
                    save('}')
            elif read == '<':
                save(string)
                string = ""
                save('<')
            elif read == '>':
                save(string)
                string = ""
                save('>')
            elif read == ',':
                save(string)
                string = ""
                save(',')
            elif read == "'":
                string += read
                if sign == 1:
                    sign = 0
                    save_const(string)
                    string = ""
                else:
                    if sign != 2:
                        sign = 1
            elif read == '"':
                string += read
                if sign == 2:
                    sign = 0
                    save_const(string)
                    string = ""
                else:
                    if sign != 1:
                        sign = 2
            elif read == ':':
                if sign == 1 or sign == 2:
                    string += read
                else:
                    save(string)
                    string = ""
                    save(':')
            elif read == '+':
                save(string)
                string = ""
                save('+')
            elif read == '=':
                save(string)
                string = ""
                save('=')
            else:
                string += read

    except Exception as e:
        print(e)


def main():
    file_name = "C:\\Data\\Projects\\PycharmProjects\\Network\\test\\test.txt"
    pretreatment(file_name)
    recognition('file.tmp')
    for i in keywords.signlist.keys():
        print("(", keywords.signlist[i], ",", i, ")")


if __name__ == '__main__':
    main()

File: src/lexer.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strings"
	"unicode"
)

var signlist = make(map[string] int, 100)

//预处理函数，将文件中的空格，换行等无关字符处理掉
func Pretreatment(fileName string) {
	//打开文件
	fpRead, err1 := os.Open(fileName)
	if !check(err1, "open") {
		return
	}
	fpWrite, err2 := os.Create("test.temp")
	if !check(err2, "open") {
		return
	}

	//关闭文件
	defer fpRead.Close()
	defer fpWrite.Close()


	sign := 0
	r := bufio.NewReader(fpRead)
	//不断读取，进行预处理
	for {
		read, err := r.ReadBytes('\n')
		//fmt.Println(err)
		if err == io.EOF {  //读文件结束
			fmt.Println("结束")
			break
		} else {
			check(err, "read")
		}

		length := len(read)
		for index := 0; index < length - 1; index++ {
			//fmt.Println(string(read[index]))
			if sign == 0 {
				if read[index] == ' ' {
					continue
				}
			}
			if read[index] == '#' { //后面均为注释，直接结束此行分析
				break
			} else if read[index] == ' ' || read[index] == '\t' || read[index] == '\n' || read[index] == '\r'{
				if sign == 1 {
					continue
				} else {
					sign = 1
					_, err := fpWrite.WriteString(" ")
					check(err, "write")
				}
			} else if read[index] == '"' || read[index] == '\'' {
				_, err1 := fpWrite.WriteString(string(read[index]))
				check(err1, "write")
				temp := read[index]
				index++
				for index < length && read[index] != temp {
					_, err := fpWrite.WriteString(string(read[index]))
					check(err, "write")
					index++
				}
				if index >= length {
					break
				}
				_, err2 := fpWrite.WriteString(string(read[index]))
				check(err2, "write")
			} else {
				sign = 3
				_, err2 := fpWrite.WriteString(string(read[index]))
				check(err2, "write")
			}
		}
	}
	//最后加一个回车，方便下次读取
	_, err := fpWrite.WriteString("\n")
	check(err, "write")
}

func save(str string) {
	//判定str是否在keywords中
	if _, ok := keywords[str]; ok {
		if _, ok := signlist[str]; !ok {
			signlist[str] = keywords[str]
		}
	} else {
		saveConst(str)
	}
}

func save_var(str string) {
	if _, ok := signlist[str]; !ok {
		if len(strings.TrimSpace(str)) >= 1 {
			if isSignal(str) == 1 {
				signlist[str] = 301
			} else {
				signlist[str] = 501
			}
		}
	}
}

func saveConst(str string) {
	if _, ok := signlist[str]; !ok {
		signlist[str] = 401
	}
}

func saveError(str string) {
	if _, ok := signlist[str]; !ok {
		signlist[str] = 501
	}
}


func isSignal(s string) int {
	if s[0] == '_' || unicode.Is(unicode.ASCII_Hex_Digit, rune(s[0])) {
		for i := 0; i < len(s); i++ {
			if unicode.Is(unicode.ASCII_Hex_Digit, rune(s[i])) || s[i] == '_' || unicode.IsDigit(rune(s[i])) {

			} else {
				return 0
			}
		}
		return 1
	} else {
		return 0
	}
}

func recognition(fileName string) {
	fpRead, err := os.Open(fileName)
	if !check(err, "open") {
		return
	}
	str := ""
	sign := 0

	r := bufio.NewReader(fpRead)
	read, err := r.ReadBytes('\n')
	//fmt.Println(err)
	if err == io.EOF {  //读文件结束
		fmt.Println("结束")
	} else {
		check(err, "read")
	}
	//对读出的一行进行处理
	for index := 0; index < len(read) - 1; index++ {
		ch := read[index]
		if ch == ' ' {
			if len(strings.TrimSpace(str)) < 1 {
				sign = 0
			} else {
				if sign == 1 || sign == 2 {
					str += string(ch)
				} else {
					save(str)
					str = ""
					sign = 0
				}
			}
		} else if ch == '(' || ch == ')' || ch == '[' || ch == ']' || ch == '{' || ch == '}' || ch == ':' {
			if sign == 1 || sign == 2 {
				str += string(ch)
			} else {
				save(str)
				str = ""
				save(string(ch))
			}
		} else if ch == '<' || ch == '>' || ch == ',' || ch == '+' || ch == '='{
			save(str)
			str = ""
			save(string(ch))
		} else if ch == '\'' {
			str += string(ch)
			if sign == 1 {
				sign = 0
				saveConst(str)
				str = ""
			} else {
				if sign != 2 {
					sign = 1
				}
			}
		} else if ch == '"' {
			str += string(ch)
			if sign == 2 {
				sign = 0
				saveConst(str)
				str = ""
			} else {
				if sign != 1 {
					sign = 2
				}
			}
		} else {
			str += string(ch)
		}
	}
}